### PR TITLE
roachtest: add pgjdbc test suite

### LINKED ISSUE #37780: roachtest: create a pgjdbc test suite roachtest
We currently have several "canary" ORM tests that run nightly. They run the test suite of the ORM or driver, using a blacklist of known-failing tests as a way to keep the test running without requiring perfect support from CockroachDB

One example is the Hibernate test, which is located in `pkg/cmd/roachtest/hibernate.go`.

We should add one for `pgjdbc`, which is the Java driver that is used to connect to Postgres.

@rafiss, I'm assigning you to take a look at doing this over time.

### PR DESCRIPTION
**Pull request by rafiss**
Also, fix instructions for generating the roachtest registry, and factor out
common code for parsing JUnit test report files.

closes #37780

Release note: None

### CODE REVIEW

**Comment by jordanlewis** [comment]
heck to the yes

**Comment by rafiss** [comment]
This doesn't include blacklists for 19.1, 2.1, or 2.0 yet. I can work on adding those if we want. (I'm not sure how the nightly roachtest works / if it will be run against those older versions.)

File: pkg/cmd/roachtest/pgjdbc_blacklist.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	{"v2.1", "pgjdbcBlackList2_1", pfjdbcBlackList2_1, "", nil},
+	{"v2.2", "pgjdbcBlackList19_1", pfjdbcBlackList19_1, "", nil},
+	{"v19.1", "pfjdbcBlackList19_1", pfjdbcBlackList19_1, "", nil},
+	{"v19.2", "pfjdbcBlackList19_2", pfjdbcBlackList19_2, "", nil},
+}
+
+// Please keep these lists alphabetized for easy diffing.
+// After a failed run, an updated version of this blacklist should be available
+// in the test log.
+var pfjdbcBlackList19_2 = blacklist{
```
Comment (jordanlewis) [inline pkg/cmd/roachtest/pgjdbc_blacklist.go]: I think this should be pgjdbc, not pfjdbc

**Review by jordanlewis — COMMENTED** [review]
I'm not sure about how the old versions get run actually. @BramGruneir do you know?

File: pkg/cmd/roachtest/pgjdbc_blacklist.go (hunk 1 above)
Comment (rafiss) [reply pkg/cmd/roachtest/pgjdbc_blacklist.go]: Yeah, that seems right. Thanks for the catch! :D

**Review by BramGruneir — CHANGES_REQUESTED** [review]
We run the older versions weekly, not nightly.  I don't know if we go all the way back to 2.0, but we were going back to then until the last release.  

Also, overall, LGTM, just a few comments.

> Reviewed 2 of 3 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-:-Ll2vAacBvmSIKiDXkRT:b-zf25j1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rafiss)

___
*[pkg/cmd/roachtest/pgjdbc.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2vuR49vZoxTpuVVhU:-Ll2vuR49vZoxTpuVVhV:b2siheb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc.go#L125)):*
> ```Go
> 		// When testing, it is helpful to run only a subset of the tests. To do so
> 		// add "-DTest=org.postgresql.package.ClassName" to the test run command.
> 		// Note that this will take upwards of 3 hours.
> ```

Are these comments still accurate?  Does it take 3 hours?
___
*[pkg/cmd/roachtest/pgjdbc.go, line 314 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2wH5wDG2Q-WuGO_g-:-Ll2wH5wDG2Q-WuGO_g0:bagat94) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc.go#L314)):*
> ```Go
> }
> 
> func extractFailureFromPgjdbcXML(contents []byte) ([]string, []bool, error) {
> ```

Is this the same as the hibernate ones?  If so, maybe make a common one for java test results or something.
___
*[pkg/cmd/roachtest/registry.go, line 16 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2wiys4eB1f7UOW_Ck:-Ll2wiys4eB1f7UOW_Cl:b-kusj6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/registry.go#L16)):*
> ```Go
> 	// Helpful shell pipeline to generate the list below:
> 	//
> 	// grep -h -E 'func register[^(]+\(.*testRegistry\) {' pkg/cmd/roachtest/*.go | grep -E -o 'register[^(]+' | grep -E -v '^register(Tests|Benchmarks)$' | grep -v '^\w*Bench$' | sort -f | awk '{printf "\t%s(r)\n", $0}'
> ```

This is a great fix, but perhaps should be in a different commit.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rafiss)

___
*[pkg/cmd/roachtest/pgjdbc.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2wMSU7vtx33OXHtUy:-Ll2wMSU7vtx33OXHtUz:b8sjbah) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc.go#L1)):*
> ```Go
> // Copyright 2018 The Cockroach Authors.
> ```

[nit]: `s/2018/2019/`.
___
*[pkg/cmd/roachtest/pgjdbc.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2wq1V4tfsf_X1cGPo:-Ll2wq1V4tfsf_X1cGPp:b-6ey2zm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc.go#L124)):*
> ```Go
> 		t.Status("running pgjdbc test suite, will take at least 3 hours")
> 		// When testing, it is helpful to run only a subset of the tests. To do so
> 		// add "-DTest=org.postgresql.package.ClassName" to the test run command.
> ```

I might be complete off-base, but maybe such a comment should be included in a "help" message through cli?
___
*[pkg/cmd/roachtest/pgjdbc_blacklist.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2x2Nz3tEsj8wrCFHW:-Ll2x2Nz3tEsj8wrCFHX:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc_blacklist.go#L1)):*
> ```Go
> // Copyright 2018 The Cockroach Authors.
> ```

ditto

**Comment by BramGruneir** [comment]
___
*[pkg/cmd/roachtest/pgjdbc.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2wq1V4tfsf_X1cGPo:-Ll2xaWC6b4VbYzv1dhy:b-pg77q0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc.go#L124)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I might be complete off-base, but maybe such a comment should be included in a "help" message through cli?
</blockquote></details>

This comment, was copy-pasted from hiberante.  It takes > 3 hours when running all the tests, so it's useful to specify a specific test by appending it to `cd /mnt/data1/pgjdbc/pgjdbc/ && mvn test` not to roachtest.

**Comment by BramGruneir** [comment]
___
*[pkg/cmd/roachtest/pgjdbc.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2wq1V4tfsf_X1cGPo:-Ll2xtLG1LQl5h7H5wAM:bl4i9tz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc.go#L124)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This comment, was copy-pasted from hiberante.  It takes > 3 hours when running all the tests, so it's useful to specify a specific test by appending it to `cd /mnt/data1/pgjdbc/pgjdbc/ && mvn test` not to roachtest.
</blockquote></details>

I don't know if pgjdbc takes as long and if this is even needed, hence my other comment.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir, @jordanlewis, and @yuzefovich)

___
*[pkg/cmd/roachtest/pgjdbc.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2wMSU7vtx33OXHtUy:-Ll30iUN-pJWBACIqkn1:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc.go#L1)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: `s/2018/2019/`.
</blockquote></details>

done
___
*[pkg/cmd/roachtest/pgjdbc.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2wq1V4tfsf_X1cGPo:-Ll3Quyu-NnJNRLlrxEP:b-762ype) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc.go#L124)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I don't know if pgjdbc takes as long and if this is even needed, hence my other comment.
</blockquote></details>

yeah,  this note probably is not needed for pgjdbc
___
*[pkg/cmd/roachtest/pgjdbc.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2vuR49vZoxTpuVVhU:-Ll30nOZ5Van7AFnt9oD:br6n5lo) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc.go#L125)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Are these comments still accurate?  Does it take 3 hours?
</blockquote></details>

thanks for the catch. i meant to adjust this after i went through a debug cycle -- no they are quite fast.
___
*[pkg/cmd/roachtest/pgjdbc.go, line 314 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2wH5wDG2Q-WuGO_g-:-Ll315ih4haZK78saWBP:b-cvexal) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc.go#L314)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Is this the same as the hibernate ones?  If so, maybe make a common one for java test results or something.
</blockquote></details>

good point, i think it's the junit result format.
___
*[pkg/cmd/roachtest/pgjdbc_blacklist.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2x2Nz3tEsj8wrCFHW:-Ll3QnzQ7gjniZDG02TT:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/pgjdbc_blacklist.go#L1)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto
</blockquote></details>

done
___
*[pkg/cmd/roachtest/registry.go, line 16 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-Ll2wiys4eB1f7UOW_Ck:-Ll3QpeABkhKn_wqNHP3:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c576d1bb1020ec2dde750c040927edddc3816fb5/pkg/cmd/roachtest/registry.go#L16)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This is a great fix, but perhaps should be in a different commit.
</blockquote></details>

done

**Review by jordanlewis — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39175#-:-LlZsqVrAW-BlMd_2mKj:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @BramGruneir, @jordanlewis, and @yuzefovich)

**Comment by rafiss** [comment]
thanks everyone

bors r+

**Pull request merged**